modules/aws-http4s/src/smithy4s/aws/internals/AwsSigning.scala

```diff
@@ -187,5 +196,39 @@ private[aws] object AwsSigning {
   private val `X-Amz-Security-Token` = CIString("X-Amz-Security-Token")
   private val `X-Amz-Target` = CIString("X-Amz-Target")
   private val algorithm = "AWS4-HMAC-SHA256"
+  private val `X-Amz-Content-SHA256` = CIString("X-Amz-Content-SHA256")
+  private val UNSIGNED_PAYLOAD = "UNSIGNED-PAYLOAD"
+
+  private sealed trait PreSigner[F[_]] {
+    def apply(request: Request[F]): F[(String, Request[F])]
+  }
+  private object PreSigner {
+    class Standard[F[_]](implicit F: Concurrent[F]) extends PreSigner[F] {
+      def apply(request: Request[F]): F[(String, Request[F])] = {
+        request.body.chunks.compile.to(Chunk).map(_.flatten).map { inMemoryBody =>
+          val payloadHash = sha256HexDigest(inMemoryBody.toArray)
+          val newRequest = request.withBodyStream(fs2.Stream.chunk(inMemoryBody))
+          (payloadHash, newRequest)
+        }
+      }
+    }
+
+    class S3Unsigned[F[_]](implicit F: Applicative[F]) extends PreSigner[F] {
+      def apply(request: Request[F]): F[(String, Request[F])] = F.pure {
+        val newRequest = request.transformHeaders(_.put(Header.Raw(`X-Amz-Content-SHA256`, UNSIGNED_PAYLOAD)))
+        (UNSIGNED_PAYLOAD, newRequest)
+      }
+    }
+
+    class S3InMemorySigned[F[_]](implicit F: Concurrent[F]) extends PreSigner[F] {
+      def apply(request: Request[F]): F[(String, Request[F])] = {
+        request.body.chunks.compile.to(Chunk).map(_.flatten).map { inMemoryBody =>
```

Comment:
**nitpick:** I don't like in this design the fact that compiling body chunks is the signer's responsibility. Maybe we should just do `.toStrict(None)` for all requests indiscriminately, and pass the result (a strict response, meaning you can consume its body N times) further in the pipeline? or even have the PreSigner receive the body as a byte chunk in addition to the request.

Reply:
I've thought about it a fair bit and I prefer for it to be the responsibility of this construct (granted, PreSigner is a shit name), because when we get around to signing streamed requests (hopefully soon), it'll have to be the responsibility of the implementation responsible of that to amend the stream of bytes to add intermediate signatures at the end of chunks, which is similar responsibility than the one held by the implementation in this PR. 

Moreover, it has strictly zero incidence in the current state, are all bodies are produced via arrays.

Reply:
ok reasonable